gdal/ogr/ogrsf_frmts/ngw/gdalngwdataset.cpp

```diff
@@ -0,0 +1,911 @@
+ */
+bool OGRNGWDataset::FlushMetadata( char **papszMetadata )
+{
+    if( !bMetadataDerty )
+    {
+        return true;
+    }
+
+    bool bResult = NGWAPI::FlushMetadata(osUrl, osResourceId, papszMetadata,
+        GetHeaders());
+    if( bResult )
+    {
+        bMetadataDerty = false;
+    }
+
+    return bResult;
+}
+
+/*
+ * SetMetadata()
+ */
+CPLErr OGRNGWDataset::SetMetadata( char **papszMetadata, const char *pszDomain)
+{
+    FetchPermissions();
+    if( !stPermissions.bMetadataCanWrite )
+    {
+        CPLError(CE_Failure, CPLE_AppDefined, "Operation not permitted.");
+        return CE_Failure;
+    }
+
+    CPLErr eResult = GDALDataset::SetMetadata(papszMetadata, pszDomain);
+    if( eResult == CE_None && EQUAL(pszDomain, "NGW") )
+    {
+        eResult = FlushMetadata( papszMetadata ) ? CE_None : CE_Failure;
+    }
+    return eResult;
+}
+
+/*
+ * SetMetadataItem()
+ */
+CPLErr OGRNGWDataset::SetMetadataItem( const char *pszName,
+    const char *pszValue, const char *pszDomain)
+{
+    FetchPermissions();
+    if( !stPermissions.bMetadataCanWrite )
+    {
+        CPLError(CE_Failure, CPLE_AppDefined, "Operation not permitted.");
+        return CE_Failure;
+    }
+    if( EQUAL(pszDomain, "NGW") )
+        bMetadataDerty = true;
+    return GDALDataset::SetMetadataItem(pszName, pszValue, pszDomain);
+}
+
+/*
+ * FlushCache()
+ */
+void OGRNGWDataset::FlushCache()
+{
+    GDALDataset::FlushCache();
+    FlushMetadata( GetMetadata("NGW") );
+}
+
+/*
+ * GetHeaders()
+ */
+char **OGRNGWDataset::GetHeaders() const
+{
+    char **papszOptions = nullptr;
+    papszOptions = CSLAddString(papszOptions, "HEADERS=Accept: */*");
+    if( !osUserPwd.empty() )
+    {
+        papszOptions = CSLAddString(papszOptions, "HTTPAUTH=BASIC");
+        std::string osUserPwdOption("USERPWD=");
+        osUserPwdOption += osUserPwd;
+        papszOptions = CSLAddString(papszOptions, osUserPwdOption.c_str());
+    }
+    return papszOptions;
+}
+
+/*
+ * ExecuteSQL()
+ */
+OGRLayer *OGRNGWDataset::ExecuteSQL( const char *pszStatement,
+    OGRGeometry *poSpatialFilter, const char *pszDialect )
+{
+    // Clean statement string.
+    CPLString osStatement(pszStatement);
+    osStatement = osStatement.Trim().replaceAll("  ", " ");
+
+    if( STARTS_WITH_CI(osStatement.c_str(), "DELLAYER:") )
+    {
+        CPLString osLayerName = osStatement.substr(9);
+        if( osLayerName.endsWith(";") )
+        {
+            osLayerName = osLayerName.substr(0, osLayerName.size() - 1);
+            osLayerName.Trim();
+        }
+
+        CPLDebug("NGW", "Delete layer with name %s.", osLayerName.c_str());
+
+        for( int iLayer = 0; iLayer < nLayers; ++iLayer )
+        {
+            if( EQUAL(papoLayers[iLayer]->GetName(), osLayerName.c_str() ) )
+            {
+                DeleteLayer( iLayer );
+                break;
+            }
+        }
+        return nullptr;
+    }
+
+    if( STARTS_WITH_CI(osStatement.c_str(), "DELETE FROM") )
```

Comment:
what about if the user issues a DELETE FROM layername WHERE ... ?

Reply:
NextGIS Web REST API only support delete specific feature or all features. So in case of `DELETE FROM layername WHERE ...`  we will get layer name like `layername WHERE ...` and GetLayerByName will not return any layer. This is expected behaviour. It will be nice to have OGRLayer::DeleteAllFeatures() method. This is an attempt to replace it. 
I will add additional error message for this case to ExecuteSQL.